File: applications/rtkimagxgeometry/rtkimagxgeometry.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkimagxgeometry_ggo.h"
#include "rtkGgoFunctions.h"
#include "rtkImagXGeometryReader.h"
#include "rtkThreeDCircularProjectionGeometryXMLFile.h"

int
main(int argc, char * argv[])
{
  GGO(rtkimagxgeometry, args_info);

  // Image Type
  constexpr unsigned int Dimension = 3;

  // Create geometry reader
  auto imagxReader = rtk::ImagXGeometryReader<itk::Image<float, Dimension>>::New();
  imagxReader->SetProjectionsFileNames(rtk::GetProjectionsFileNamesFromGgo(args_info));
  if (args_info.calibration_given)
  {
    imagxReader->SetCalibrationXMLFileName(args_info.calibration_arg);
  }
  if (args_info.room_setup_given)
  {
    imagxReader->SetRoomXMLFileName(args_info.room_setup_arg);
  }
  TRY_AND_EXIT_ON_ITK_EXCEPTION(imagxReader->UpdateOutputData())

  // Write
  TRY_AND_EXIT_ON_ITK_EXCEPTION(rtk::WriteGeometry(imagxReader->GetGeometry(), args_info.output_arg))

  return EXIT_SUCCESS;
}

File: applications/rtkimagxgeometry/rtkimagxgeometry.py
#!/usr/bin/env python
import argparse
import itk
from itk import RTK as rtk


def build_parser():
    parser = rtk.RTKArgumentParser(
        description="Creates an RTK geometry file from an iMagX acquisition."
    )

    parser.add_argument("--calibration", "-c", help="iMagX Calibration file")
    parser.add_argument("--room_setup", "-s", help="iMagX room setup file")
    parser.add_argument("--output", "-o", required=True, help="Output file name")
    parser.add_argument(
        "--verbose", "-v", action="store_true", help="Verbose execution"
    )

    # Projections group (path, regexp, nsort, submatch)
    rtk.add_rtkinputprojections_group(parser)

    return parser


def process(args_info: argparse.Namespace):
    Dimension = 3

    # Create geometry reader
    imagxReader = rtk.ImagXGeometryReader[itk.Image[itk.F, Dimension]].New()
    fileNames = rtk.GetProjectionsFileNamesFromArgParse(args_info)
    imagxReader.SetProjectionsFileNames(fileNames)

    if getattr(args_info, "calibration", None):
        imagxReader.SetCalibrationXMLFileName(args_info.calibration)
    if getattr(args_info, "room_setup", None):
        imagxReader.SetRoomXMLFileName(args_info.room_setup)

    imagxReader.UpdateOutputData()
    # Write
    rtk.WriteGeometry(imagxReader.GetGeometry(), args_info.output)

    return 0


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    return process(args_info)


if __name__ == "__main__":
    main()
